Detect keywords when those were not provided as params

petrican opened the pull request:
<!--
Thank you for your pull request! 

Please title your pull request like this: 'Module: Change', eg 'Fraggles Bid Adapter: support fragglerock'

Please make sure this PR is scoped to one change or you may be asked to resubmit. 
 
Please make sure any added or changed code includes tests with greater than 80% code coverage. 

See https://github.com/prebid/Prebid.js/blob/master/CONTRIBUTING.md#testing-prebidjs for documentation on testing Prebid.js.

For any user facing change, submit a link to a PR on the docs repo at https://github.com/prebid/prebid.github.io/
-->

## Type of change
<!-- Remove items that don't apply and/or select an item by changing [ ] to [x] -->
- [x] Bugfix
- [ ] Feature
- [ ] New bidder adapter  <!--  IMPORTANT: also submit your bidder parameter documentation as noted in https://docs.prebid.org/dev-docs/bidder-adaptor.html#submitting-your-adapter -->
- [ ] Updated bidder adapter  <!--  IMPORTANT: (1) consider whether you need to upgrade your bidder parameter documentation in https://github.com/prebid/prebid.github.io/tree/master/dev-docs/bidders and (2) if you have a Prebid Server adapter, please consider whether that should be updated as well. --> 
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes

- [ ] Does this change affect user-facing APIs or examples documented on http://prebid.org?
- [ ] Other

## Description of change
<!-- Describe the change proposed in this pull request -->
Detect the keywords even if those weren't provided by the user in the config params


<!-- For new bidder adapters, please provide the following
- contact email of the adapter’s maintainer
- test parameters for validating bids:
```
{
  bidder: '<bidder name>',
  params: {
    // ...
  }
}
```

Be sure to test the integration with your adserver using the [Hello World](https://github.com/prebid/Prebid.js/blob/master/integrationExamples/gpt/hello_world.html) sample page. -->


## Other information
<!-- References to related PR or issue #s, @mentions of the person or team responsible for reviewing changes, etc. -->

petrican commented:
Hi! Can anyone help me with an input/review here too please ? Thanks!

petrican commented:
> hey, i noticed you have no tests and pulling keywords already exists a bit in core. how is this different?

Hi @patmmccann! Thank you for your review. I added tests for each of the cases now. We added these functions as we need to extract the keywords when they are not provided via ortb. If you are looking closely to the existing functions from the lib https://github.com/Sevio/Prebid.js/blob/master/libraries/keywords/keywords.js those rely on the user to provide those. But if those aren't provided we would like a way to fallback and try the detection routines I added to the adapter. Basically the functions try to pull keywords from the page when possible, but if the site doesn’t include any (no meta, no JSON-LD, no meaningful title/H1), then it falls back to whatever the bidder or user explicitly provides. Let me know please if there is something else I should add.

patmmccann commented:
don't we get them here already? https://github.com/prebid/Prebid.js/blob/56a41d89c389e0695186b9da3160c8914ec044e4/src/fpd/enrichment.ts#L167

petrican commented on modules/sevioBidAdapter.js:
```diff
@@ -9,6 +9,170 @@ const GVLID = `1393`;
+  return keywords;
+};
+
+const detectHeaderKeywords = () => {
```
The enrichment code only copies from <meta name="keywords"> when it can read it (same-origin, tag present before requestBids) as I understand it, so in that case nothing is going to be added as keywords. Supplying ortb2.site.keywords via config solves this. We wanted a way to still get the keywords to be able to provide more targeted ads to the end user.

petrican commented:
> still appears to be redundant with core. In general I am quite skeptical of this functionality being in a bid adapter instead of somewhere accessible by all. What are your goals here?

Hi @patmmccann ! We just want a 'fallback' way to have keywords sent whenever possible within the publisher's site to deliver meaningful ads. Thanks

petrican commented:
I have modified the code to use the keywords it detects from the ortb. I also removed the functions I wrote for custom detection since I understand that is the general desire to use and rely the core existing functions. Please check if this is fine now. Thanks!

petrican commented:
I have added the tests requested. Thanks!
